# 实验报告 #1

> 姓名：林佳胜      学号：PB24511997
>
> 时间：2026 年 3 月 4 日

---

## 实验目的

本实验围绕三类典型数值计算问题展开：  

1. 两相近数相减；
2. 大数吃小数；
3. 函数近似算法精度。  

目标是在相同数学问题下比较不同实现方式的数值稳定性，并给出更可靠的计算策略。

---

## 运行环境

- 操作系统：`Windows 10`
- 编程语言：`Python 3.11.9`
- 主要依赖库：`numpy>=2.4.2`

---


## 实验内容与结果

### 问题 1

给定两个函数：
$$f(x)=\sqrt{x^2+81}-9 \quad \text{和} \quad g(x)=\frac{x^2}{\sqrt{x^2+81}+9}$$

**编程要求**：采用**单精度 (float32)** 进行编程。注意：开方 $\sqrt{x}$ 等内置函数的输出结果默认是双精度 (float64)，**需要强制转成单精度**。

**数据要求**：分别取 $x=4^{-1}, 4^{-2}, 4^{-3}, \cdots, 4^{-11}$，输出相应的函数值 $f(x)$ 和 $g(x)$。

**格式要求**：计算结果**保留 12 位尾数**（用科学计数形式）。

**分析要求**：比较并分析两种方法得到的计算结果。你认为哪种方法得到的计算结果更可靠？请给出你的理由或分析。

In [1]:
import numpy as np

# 类型说明
Float32 = np.float32
ArrF32 = np.typing.NDArray[Float32]

# 函数定义
def f32(x: Float32) -> Float32:
    temp1: Float32 = x * x + Float32(81.0)
    temp2: Float32 = Float32(np.sqrt(temp1))
    return temp2 - Float32(9.0)

def g32(x: Float32) -> Float32:
    temp1: Float32 = x * x
    temp2: Float32 = temp1 + Float32(81.0)
    temp3: Float32 = Float32(np.sqrt(temp2)) + Float32(9.0)
    return temp1 / temp3

# 输入准备
x_array: ArrF32 = [Float32(4 ** (-x)) for x in range(1, 12)]

# 输出设置
print(f"{'id':>3}\t{'x':>18}\t{'f(x)':>18}\t{'g(x)':>18}")
print("=" * 75)
for idx, x in enumerate(x_array, start=1):
    f_x: Float32 = f32(x)
    g_x: Float32 = g32(x)
    print(f"{idx:3d}\t{Float32(x):18.12E}\t{Float32(f_x):18.12E}\t{Float32(g_x):18.12E}")

 id	                 x	              f(x)	              g(x)
  1	2.500000000000E-01	3.471374511719E-03	3.471552627161E-03
  2	6.250000000000E-02	2.174377441406E-04	2.170112711610E-04
  3	1.562500000000E-02	1.335144042969E-05	1.356335815217E-05
  4	3.906250000000E-03	9.536743164062E-07	8.477105097882E-07
  5	9.765625000000E-04	0.000000000000E+00	5.298190686176E-08
  6	2.441406250000E-04	0.000000000000E+00	3.311369178860E-09
  7	6.103515625000E-05	0.000000000000E+00	2.069605736787E-10
  8	1.525878906250E-05	0.000000000000E+00	1.293503585492E-11
  9	3.814697265625E-06	0.000000000000E+00	8.084397409326E-13
 10	9.536743164062E-07	0.000000000000E+00	5.052748380829E-14
 11	2.384185791016E-07	0.000000000000E+00	3.157967738018E-15


#### 实验结果与分析 (问题一)

| $x$ | $f(x) = \sqrt{x²+81}-9$ | $g(x) = \frac{x^2}{\sqrt{x²+81}+9}$ |
|:---:|---|---|
| $4^{-1}$ | 3.471374511719E-03 | 3.471552627161E-03 |
| $4^{-2}$ | 2.174377441406E-04 | 2.170112711610E-04 |
| $4^{-3}$ | 1.335144042969E-05 | 1.356335815217E-05 |
| $4^{-4}$ | 9.536743164062E-07 | 8.477105097882E-07 |
| $4^{-5}$ | 0.000000000000E+00 | 5.298190686176E-08 |
| $4^{-6}$ | 0.000000000000E+00 | 3.311369178860E-09 |
| $4^{-7}$ | 0.000000000000E+00 | 2.069605736787E-10 |
| $4^{-8}$ | 0.000000000000E+00 | 1.293503585492E-11 |
| $4^{-9}$ | 0.000000000000E+00 | 8.084397409326E-13 |
| $4^{-10}$ | 0.000000000000E+00 | 5.052748380829E-14 |
| $4^{-11}$ | 0.000000000000E+00 | 3.157967738018E-15 |

**分析：**

当 $x$ 很小时，函数 $f(x)$ 的计算结果趋近于零，但由于数值精度的限制，最终结果被截断为零。这是因为在计算 $\sqrt{x^2+81}-9$ 时，两个数值非常接近，导致了**严重的数值不稳定性**（即所谓的“相近数相减”问题）。这种情况下，计算机无法准确表示结果，导致了精度丢失。

**结论：**

方法 $g(x)$ 得到的计算结果更可靠。虽然两者在数学上是等价的，但 $g(x)$ 避免了“相近数相减问题”，从而提高了算法的数值稳定性。

---

### 问题 2

给定如下数据：

$$4044.045051380452, \quad 0.000531415926535, \quad -2759471.276702747,$$
$$0.0000557052996742895, \quad 2755463.874010974, \quad -36.64291531256604,$$
$$-0.000031415926535$$

分别采取以下 4 种方式求和：

**(a) 顺序求和**；

**(b) 逆序（从后往前）求和**；

**(c) 按绝对值从小到大的顺序**，依次求和；

**(d) 按绝对值从大到小的顺序**，依次求和。

采用**双精度（float64）**进行计算，计算结果中的**尾数至少保留 9 位小数**（用科学计数形式，例如 1.234567899E-11）。

**分析要求**：比较 4 种方法得到的计算结果；你认为哪种方法得到的计算结果更精确（即误差最小；**提示：想办法算出精确值**）？试给出你的理由或分析。

In [2]:
import numpy as np
from decimal import Decimal

# 类型说明
Float64 = np.float64
ArrF64 = np.typing.NDArray[Float64]

# 函数定义
# a. 顺序求和
def sum_forward(arr: ArrF64) -> Float64:
    result: Float64 = Float64(0.0)
    for val in arr:
        result += val
    return result

# b. 逆序求和
def sum_backward(arr: ArrF64) -> Float64:
    result: Float64 = Float64(0.0)
    for val in reversed(arr):
        result += val
    return result

# c. 按绝对值从小到大求和
def sum_small_to_large(arr: ArrF64) -> Float64:
    result: Float64 = Float64(0.0)
    for val in arr[np.argsort(np.abs(arr))]:
        result += val
    return result

# d. 按绝对值从大到小求和
def sum_large_to_small(arr: ArrF64) -> Float64:
    result: Float64 = Float64(0.0)
    for val in arr[np.argsort(-np.abs(arr))]:
        result += val
    return result

# 输入准备
data: ArrF64 = np.array([
    4044.045051380452, 
    0.000531415926535, 
    -2759471.276702747, 
    0.0000557052996742895,
    2755463.874010974, 
    -36.64291531256604, 
    -0.000031415926535], dtype=Float64)

# 输出准备
precise_value: Float64 = Float64(sum(Decimal(str(x)) for x in data))
result_a = sum_forward(data)
result_b = sum_backward(data)
result_c = sum_small_to_large(data)
result_d = sum_large_to_small(data)

# 输出结果
print(f"{'Method':>15}\t\t{'result':>20}\t\t{'Absolute error (abs)':>20}")
print("=" * 77)
print(f"{'Exact value':>15}\t\t{precise_value:20.12E}\t\t{'':>20}") 
print(f"{'Method a':>15}\t\t{result_a:20.12E}\t\t{abs(result_a - precise_value):20.12E}")
print(f"{'Method b':>15}\t\t{result_b:20.12E}\t\t{abs(result_b - precise_value):20.12E}")
print(f"{'Method c':>15}\t\t{result_c:20.12E}\t\t{abs(result_c - precise_value):20.12E}")
print(f"{'Method d':>15}\t\t{result_d:20.12E}\t\t{abs(result_d - precise_value):20.12E}")

         Method		              result		Absolute error (abs)
    Exact value		  1.856342895000E-10		                    
       Method a		  1.262001062056E-11		  1.730142788794E-10
       Method b		 -6.821210263297E-11		  2.538463921330E-10
       Method c		  0.000000000000E+00		  1.856342895000E-10
       Method d		  4.684540313304E-11		  1.387888863670E-10


#### 实验结果与分析 (问题二)

从计算结果可以看出，四种求和方式得到的结果差异很大：

| 方法                 | 求和结果              | 绝对误差大小         |
|:-------------------:|:---------------------:|:------------------:|
| 精确值              | 1.856342895000E-10    | -                  |
| 顺序求和 (a)        | 1.262001062056E-11    | 1.730142788794E-10 |
| 逆序求和 (b)        | -6.821210263297E-11   | 2.538463921330E-10 |
| 绝对值从小到大 (c)   | 0.000000000000E+00    | 1.856342895000E-10 |
| 绝对值从大到小 (d)   | 4.684540313304E-11    | 1.387888863670E-10 |

**分析：**

1.  **“大数吃小数”现象**：当绝对值很大的数（如 $10^6$ 量级）与绝对值很小的数（如 $10^{-5}$ 量级）直接相加时，由于双精度浮点数只有约 16 位有效数字，小数部分的低位信息会被舍入掉。
2.  **方法 (c) 的失效**：虽然通常建议“从小到大”求和以减少误差，但在本题中，数据包含正负抵消。方法 (c) 先将所有小数累加，但在与后续的大数运算时，累加的小数和依然远小于大数，导致信息在最后阶段被全部抹除（结果为 0）。
3.  **方法 (d) 的相对优越性**：在本组具有**两大数相近且符号相反**特性的数据中，先让大数相互抵消（$-2.759\times 10^6$ 与 $2.755\times 10^6$ 相加），减小了中间结果的量级，使后续的小数能够保留更多的有效位。

**结论：**

在数值计算中，求和顺序显著影响结果精度。对于包含正负抵消的序列，**按绝对值从大到小 (d)** 运算大概率更精确，因为它能迅速降低中间结果的量级，减轻“大数吃小数”的程度。


---

### 问题 3

**（计算结果保留 10 位有效数字）**

设 $\pi \approx 3.1415926535897932$。

使用特殊的 $\sin(x)$ 计算方法（如下所示）来计算 $\sin(x)$。

**特殊计算方法**：对于 $x$ 的不同取值范围，采用以下策略：

1. 若 $|x| < 10^{-8}$，则 $\sin(x) \approx x$

2. 若 $|x| > \frac{\pi}{6}$，利用三倍角公式 $\sin(3u) = 3\sin(u) - 4\sin^3(u)$，令 $u = \frac{x}{3}$，计算 $\sin(u)$，再由三倍角公式得到 $\sin(x)$

3. 若 $|x| \leq \frac{\pi}{6}$，令 $u = x$ ，并使用如下方式进行计算：
$$\sin(u) = u \cdot \frac{1 - \frac{29593}{207636}u^2 + \frac{34911}{7613320}u^4 - \frac{479249}{11511339840}u^6}{1 + \frac{1671}{69212}u^2 + \frac{97}{351384}u^4 + \frac{2623}{1644477120}u^6}$$

**要求**：

分别取 $x = \frac{\pi}{2024}, \frac{\pi}{1012}, \frac{\pi}{50}, \frac{\pi}{8}, \frac{\pi}{6}, \frac{\pi}{4}, \frac{\pi}{3}$，用上述特殊方法计算 $\sin(x)$。

同时，使用 **Taylor 级数的前 4 项**逼近：
$$\sin(x) \approx x - \frac{x^3}{3!} + \frac{x^5}{5!} - \frac{x^7}{7!}$$

**对比分析**：比较这两种方法得到的计算结果，分析它们在不同 $x$ 值范围内的精度差异。

In [3]:
import math

# 常数定义
PI = 3.1415926535897932

# 辅助函数
def reduce_range(x: float) -> float:
    """将 x 的范围限定为 [-π/2, π/2]。"""
    
    # 此步会将 x 缩放到 [-π, π]
    x = math.remainder(x, 2.0 * PI)  
    
    if x > PI / 2:
        x = PI - x
    elif x < -PI / 2:
        x = -PI - x

    return x

def sin_special_helper(u: float) -> float:
    """使用给定的公式计算 sin u 。"""
    u2 = u * u
    u4 = u2 * u2
    u6 = u2 * u4

    # 分子计算
    num: float = (
        1 - (29593 / 207636) * u2 
        + (34911 / 7613320) * u4 
        - (479249 / 11511339840) * u6
    )

    # 分母计算 
    den: float = (
        1 + (1671 / 69212) * u2 
        + (97 / 351384) * u4
        + (2623 / 1644477120) * u6
    )

    return u * (num / den)

# 1. 特殊 sin
def sin_special(x: float) -> float:
    x = reduce_range(x)

    # a. |x| < 10^-8
    if abs(x) < 1e-8:
        return x

    # b. |x| > π/6
    if abs(x) > PI / 6:
        u = x / 3
        sin_u = sin_special_helper(u)
        sin_x = (3 - 4 * sin_u * sin_u) * sin_u

    else:
        sin_x = sin_special_helper(x)
    
    return sin_x

# 2. Taylor 展开
def sin_taylor(x: float) -> float:
    x = reduce_range(x)

    x2 = x * x
    x3 = x2 * x
    x5 = x2 * x3
    x7 = x5 * x2

    return x - x3 / 6. + x5 / 120. - x7 / 5040.

# 数据准备
x_array = [
    ("π/2024", PI / 2024),
    ("π/1012", PI / 1012),
    ("π/50", PI / 50),
    ("π/8", PI / 8),
    ("π/6", PI / 6),
    ("π/4", PI / 4),
    ("π/3", PI / 3),
]

# 输出结果
print(f"{'x':>8}    {'Special Routine':>16}\t{'Taylor 4-term':>16}\t{'Precise':>16}\t {'Abs_err_sp(%)':>14}\t{'Abs_err_ta(%)':>14}")
print("=" * 111)
for label, x in x_array:
    special = sin_special(x)
    taylor = sin_taylor(x)
    precise = math.sin(x)

    err_sp = abs(special - precise) / precise * 100
    err_ta = abs(taylor - precise) / precise * 100
    print(f"{label:>8}    {special:16.10E}\t{taylor:16.10E}\t{precise:16.10E}\t{err_sp:14.10f}\t{err_ta:14.10f}")

       x     Special Routine	   Taylor 4-term	         Precise	  Abs_err_sp(%)	 Abs_err_ta(%)
  π/2024    1.5521696601E-03	1.5521696601E-03	1.5521696601E-03	  0.0000000000	  0.0000000000
  π/1012    3.1043355807E-03	3.1043355807E-03	3.1043355807E-03	  0.0000000000	  0.0000000000
    π/50    6.2790519529E-02	6.2790519529E-02	6.2790519529E-02	  0.0000000000	  0.0000000000
     π/8    3.8268343237E-01	3.8268343175E-01	3.8268343237E-01	  0.0000000000	  0.0000001597
     π/6    5.0000000000E-01	4.9999999187E-01	5.0000000000E-01	  0.0000000000	  0.0000016262
     π/4    7.0710678119E-01	7.0710646958E-01	7.0710678119E-01	  0.0000000000	  0.0000440685
     π/3    8.6602540378E-01	8.6602127166E-01	8.6602540378E-01	  0.0000000000	  0.0004771371


#### 实验结果与分析 (问题三)

计算特殊 sin 计算方法与 Taylor 4 项展开式的结果对比：

| $x$ | 特殊计算方法 | Taylor 4项展开 | 参考值 `math.sin(x)` | 特殊法相对误差大小(%) | Taylor相对误差大小(%) |
|:---:|---:|---:|---:|---:|---:|
| $\frac{π}{2024}$ | 1.5521696601E-03 | 1.5521696601E-03 | 1.5521696601E-03 | 0.0000000000 | 0.0000000000 |
| $\frac{π}{1012}$ | 3.1043355807E-03 | 3.1043355807E-03 | 3.1043355807E-03 | 0.0000000000 | 0.0000000000 |
| $\frac{π}{50}$ | 6.2790519529E-02 | 6.2790519529E-02 | 6.2790519529E-02 | 0.0000000000 | 0.0000000000 |
| $\frac{π}{8}$ | 3.8268343237E-01 | 3.8268343175E-01 | 3.8268343237E-01 | 0.0000000000 | 0.0000001597 |
| $\frac{π}{6}$ | 5.0000000000E-01 | 4.9999999187E-01 | 5.0000000000E-01 | 0.0000000000 | 0.0000016262 |
| $\frac{π}{4}$ | 7.0710678119E-01 | 7.0710646958E-01 | 7.0710678119E-01 | 0.0000000000 | 0.0000440685 |
| $\frac{π}{3}$ | 8.6602540378E-01 | 8.6602127166E-01 | 8.6602540378E-01 | 0.0000000000 | 0.0004771371 |

**分析：**

1. **小角度范围内两者表现相当**：
   
   当 $x \leq \frac{\pi}{50}$ 时，两种方法的相对误差均在 $10^{-10}$ 以下，难以区分。这是因为在小角度时，Taylor 级数的前 4 项已经足够精确。

2. **角度增大时 Taylor 误差快速增长**：
   - 在 $x = \frac{\pi}{8}$ 时，Taylor 误差达到 $1.6 \times 10^{-7}$（%）；
   - 在 $x = \frac{\pi}{3}$ 时，Taylor 误差达到 $4.8 \times 10^{-4}$（%），约增长 3000 倍。
   
   这表明 Taylor 级数的收敛性随角度增大而迅速恶化。

3. **特殊方法始终维持极高精度**：
   
   在全范围内相对误差均 $< 10^{-10}$，表现基本依赖于双精度浮点数的舍入精度，与 $x$ 的大小关系不大。

4. **方法原理对比**：
   - **特殊方法**：组合使用了三倍角公式和有理函数逼近，避免了高次幂的直接运算；
   - **Taylor 4 项展开**：在 $|x|$ 接近 $\frac{\pi}{3}$ 时高次项（$x^7$）不可忽略，导致误差累积。

**结论：**

对于 $|x| > \frac{\pi}{6}$ 的场景，特殊方法的精度远超 Taylor 4 项，体现了算法选择对数值稳定性的重要影响。

---


## 算法分析（综合）

本实验通过三个典型案例揭示了数值计算中的核心问题及其解决策略：

### 1. 相近数相减的数值不稳定性

**现象**：当 $x < 4^{-4}$ 时，$f(x) = \sqrt{x^2+81}-9$ 直接计算结果全部归零。

**根源**：两个相近的大数相减（$\sqrt{x^2+81} \approx 9$）导致有效数字大量损失。单精度浮点数仅 6-7 位有效数字，减法后所剩无几。

**解决方案**：通过代数变换 $g(x) = \frac{x^2}{\sqrt{x^2+81}+9}$ 将减法转化为除法，避免直接相减。

**启示**：在数值计算中应尽量**避免量级相近的数值直接相减**。

---

### 2. 大数吃小数的累积效应

**现象**：四种求和顺序产生了 $10^{-11}$ 到 $10^{-10}$ 量级的巨大差异，相对误差最高达 137%。

**根源**：浮点数加法时，量级相差超过“双精度有效位数”的小数会被直接舍弃。

**关键发现**：
- 方法 (c) 在本题中**完全失效**（结果为 0），因为小数累加后仍远小于大数；
- 方法 (d) 表现最优，因为它优先让 $\pm 10^6$ 量级的大数相互抵消，降低中间结果量级。

**启示**：求和顺序选择需**结合数据特征**：
- 若无正负抵消，从小到大一般是最优的；
- 若存在大数相互抵消，应**优先处理大数**，再累加小数。

---

### 3. 算法设计对精度的决定性影响

**现象**：特殊方法在全范围保持 $< 10^{-10}$ 的误差，而 Taylor 4 项在 $x = \frac{\pi}{3}$ 时误差达 $4.8 \times 10^{-4}\%$。

**对比分析**：

| 特性 | 特殊方法 | Taylor 4 项 |
|:---:|:---:|:---:|
| **核心策略** | 三倍角递推 + 有理函数逼近 | 幂级数截断 |
| **适用范围** | 全区间高精度 | 仅小角度（$x < 0.1$）精确 |
| **误差增长** | 几乎不变 | 随 $x$ 增大呈指数增长 |

**启示**：

- 三角函数的高精度实现应采用“**区间缩减 + 分段近似**”框架，而不是在全区间直接套同一近似式。  
- 固定阶数的 Taylor 展开只在小角度有效；当输入变大时，应通过变换（如三倍角/半角）先把自变量降到稳定区间。  
- 算法设计应在满足精度的前提下，再综合考虑计算量与实现复杂度。  

---

## 实验结论

1. **数值稳定性比公式等价更重要**：
   
   等价公式在浮点计算中的误差可能差异很大。

2. **误差与运算方式强相关**：
   
   相近数相减会丢失有效位，大数与小数混合求和会产生舍入误差。

3. **算法选择决定精度上限**：
   
   小角度下 Taylor 近似可用；在更大区间，分段策略（三倍角 + 有理逼近）更稳定。

**总结**：应根据数据特征与精度目标选择算法，而不是只依据数学形式。